loader/loader.go

```diff
@@ -1188,110 +1188,367 @@ func fetchMatchedLiteral(ctx *tcontext.Context, router *router.Table, schema, ta
 	return targetSchema, targetTable
 }
 
-func (l *Loader) restoreData(ctx context.Context) error {
-	begin := time.Now()
+// restore schema job
+type restoreSchemaJob struct {
+	loader           *Loader
+	session          *DBConn
+	database         string
+	table            string // empty if it's a schema of database
+	pathOfSchemaFile string
+}
 
-	baseConn, err := l.toDB.GetBaseConn(ctx)
-	if err != nil {
-		return err
+// check is schema of database
+func (job *restoreSchemaJob) isSchemaOfDatabase() (yes bool) {
+	yes = (job.table == "")
+	return
+}
+
+// check is schema of table
+func (job *restoreSchemaJob) isSchemaOfTable() (yes bool) {
+	yes = (job.table != "")
+	return
+}
+
+// job queue of schema restoring
+type jobQueue struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	msgq   chan *restoreSchemaJob
+	closed bool         // msgq closed flag
+	lock   sync.RWMutex // rwlock of `closed`
+	err    chan error
+}
+
+// `newJobQueue` consturct a jobQueue
+func newJobQueue(ctx context.Context) *jobQueue {
+	selfCtx, cancel := context.WithCancel(ctx)
+	return &jobQueue{
+		ctx:    selfCtx,
+		cancel: cancel,
+		msgq:   make(chan *restoreSchemaJob, 1),
+		err:    make(chan error),
+		closed: false,
 	}
-	defer func() {
-		err2 := l.toDB.CloseBaseConn(baseConn)
-		if err2 != nil {
-			l.logger.Warn("fail to close connection", zap.Error(err2))
+}
+
+// broke job queue
+// -> err
+// <- done
+func (q *jobQueue) broke(err error) {
```

Comment:
```suggestion
func (q *jobQueue) break(err error) {
```

Reply:
I think we could use `errgroup` functional replace err channel relative block of code to decrease code complexity. How about it?

Reply:
I just realized that `errgroup.Group` maybe is not suitable for this case. we run multiple consumers concurrently, at mean time we are watching error channel when produce new jobs. If we use `errgroup.Group`, `push` method may look like that:

```go
errGroup.Go(func () error {
   errCh := make(chan error)
   go func () {
       errCh  <- errGroup.Wait() 
   }()
    select {
       case err  := <- ctx.Done():
           return err
       case err := <- errCh
           return err
       case  msgq <- job:
           return nil
    }
})
```

hmmm, what do you think?

Reply:
I will push an update commit(use `errgroup.Group`) today, then we could take a look.